Netperf format change pr

noahfalk opened the pull request:
I am still working to finish the file format changes and merge with Jose's session changes, but I wanted to start the pipeline of PR feedback. Depending on amount of feedback and CI overhead I could either add more work into this PR or check it in and do additional work in a new PR. Although the work might get large in total each commit should represent a modest change with a specific goal that is reviewable in isolation. Hopefully that aids in review.

These first few commits do some preparatory work and address issue #19688. Once the full stream of work is complete I also aim to address issues #23414, #24188, #20751, #20555, #21827, #24852, and #25046


@janvorli - Could you take a look at the 64 bit thread support in step 3?
@dotnet/dotnet-diag

[EDIT]: Adding a fix for #25046

mjsabby commented:
@noahfalk If we're going this far, we should add a field to indicate if the eventblock is compressed or not. Or better yet a compressionId, and our documentation states what the id means in terms of what compression was done.

janvorli commented:
@noahfalk The 64 bit thread id support looks fine.

noahfalk commented:
@msjabby - If the header compression (last bullet in #24852) pans out I expect to have a byte in the block header which toggles between 0 and 1 to indicate whether that compression is in effect. I'm using it as a flag but you could imagine adding more values to the enumeration in some future revision of the file format.

noahfalk commented:
@janvorli - Thanks!

noahfalk commented:
@dotnet/dotnet-diag I just pushed the rest of the changes. Sorry for doing a big PR but without the recent talk about CI instability I didn't want to risk the overhead of multiple smaller PRs. The changes are divided up into commits and you can think of each commit being its own PR for one unit of work. I'm still testing them and I know there are compile issues on Linux, so there may be a small amount of cleanup but I am hoping no further significant changes in this PR.

The known work still to come (in a different PR) would be:
1) updating the configuration so the new format is easier to select via IPC / make it default
2) Maoni's request for processor number
3) I did also notice there is a pre-existing issue where threads can pile up indefinitely in buffer manager lists which made it problematic to send thread-exit notifications within the file format.

jorive commented on src/vm/eventpipebuffer.cpp:
```diff
@@ -144,65 +149,49 @@ EventPipeEventInstance *EventPipeBuffer::GetNext(EventPipeEventInstance *pEvent,
         }
 
-        if (pEvent->GetData())
+        if (m_pCurrentReadEvent->GetData())
```
In the case you set this pointer to NULL in if block above, this will AV.

jorive replied:
Would it make more sense to create a
```cpp
EventPipeEventInstance *GetNextAlignedAddress(EventPipeEventInstance *pEventInstance)
{
    const auto Offset = m_pCurrentReadEvent->GetDataLength() ? m_pCurrentReadEvent->GetDataLength() : 1;
    return const_cast<BYTE *>(m_pCurrentReadEvent->GetData() + Offset);
}
```

noahfalk replied:
AV - yep that is bad : ) The code shouldn't ever go down that path but no reason to live so dangerously.

Refactoring-wise I'm going to leave this as-is unless you feel strongly about it. Certainly it could also be written with the ternary operator, but the line would get pretty long or you would need to split it into multiple lines. The expression above isn't quite the right calculation, I think it would need to be something like:

```
EventPipeEventInstance *GetNextAlignedAddress(EventPipeEventInstance *pEventInstance)
{
    const BYTE* pUnalignedAddress = m_pCurrentReadEvent->GetDataLength() ? 
        const_cast<BYTE*>(m_pCurrentReadEvent->GetData() + m_pCurrentReadEvent->GetDataLength()) : 
        const_cast<BYTE*>(m_pCurrentReadEvent + 1);
    return (EventPipeInstance*)GetNextAlignedAddress(pUnalignedAddress);
}
```

Extracting that out as a function might make the code in this function a little easier to follow, but IMO it felt like too small of an improvement not to give deference to the pre-existing code structure.

jorive commented on src/vm/eventpipebuffer.cpp:
```diff
@@ -144,65 +149,49 @@ EventPipeEventInstance *EventPipeBuffer::GetNext(EventPipeEventInstance *pEvent,
     {
         // Confirm that pEvent is within the used range of the buffer.
-        if (((BYTE *)pEvent < m_pBuffer) || ((BYTE *)pEvent >= m_pCurrent))
+        if (((BYTE*)m_pCurrentReadEvent < m_pBuffer) || ((BYTE*)m_pCurrentReadEvent >= m_pCurrent))
```
Is this check meant to be outside this if-block?

noahfalk replied:
I think its correct where it is, was there anything specific that looked bad? (Aside from that AV below which is definately a bug : )

jorive replied:
The comment was about the AV below.

jorive commented on src/vm/eventpipebuffer.cpp:
```diff
@@ -213,23 +202,28 @@ EventPipeEventInstance *EventPipeBuffer::PeekNext(LARGE_INTEGER beforeTimeStamp)
 
-    if (pNext != NULL)
-        m_pLastPoppedEvent = pNext;
+    _ASSERTE(m_state == READ_ONLY);
```
nit: move to `CONTRACT` under `PRECONDITION`

jorive commented:
>         POSTCONDITION(RETVAL % ALIGNMENT_SIZE == 0);

Maybe we could add a PRECONDITION on format?

---
Refers to: src/vm/eventpipeeventinstance.cpp:77 in 929ed26. [](commit_id = 929ed26814f9877a5648b8cac9e275b7d9e49a47, deletion_comment = False)

jorive commented on src/vm/eventpipeeventinstance.cpp:
```diff
@@ -79,16 +79,35 @@ unsigned int EventPipeEventInstance::GetAlignedTotalSize() const
+            m_dataLength +                  // Event payload data
+            sizeof(unsigned int) +          // Prepended stack payload size in bytes
+            m_stackContents.GetSize();      // Stack payload size
+    }
```
Maybe it should be like this:
```cpp
    unsigned int threadIdSize = 0;

    if (format == EventPipeNetPerfFormatV3)
    {
        threadIdSize = sizeof(DWORD);
    }
    else if (format == EventPipeNetTraceFormatV4)
    {
        threadIdSize = sizeof(m_threadId);
    }
    else
    {
        // Error out?
    }

    payloadLength =
        sizeof(m_metadataId) +         // Metadata ID
        threadIdSize +                          // Thread ID
        sizeof(m_timeStamp) +          // TimeStamp
        sizeof(m_activityId) +             // Activity ID
        sizeof(m_relatedActivityId) + // Related Activity ID
        sizeof(m_dataLength) +          // Data payload length
        m_dataLength +                      // Event payload data
        sizeof(unsigned int) +             // Prepended stack payload size in bytes
        m_stackContents.GetSize();    // Stack payload size
```

noahfalk replied:
I added the error out part, but I left the sums separate because several more changes accumulate over the coming commits.

jorive commented on src/vm/eventpipesession.cpp:
```diff
@@ -319,7 +338,8 @@ void EventPipeSession::WriteEvent(EventPipeEventInstance &instance)
         return;
-    m_pFile->WriteEvent(instance);
+    EventPipeSequencePoint sequencePoint;
+    m_pFile->WriteSequencePoint(&sequencePoint);
```
nit: Use reference, or add PRECONDITION and if (pSequencePoint == nullptr) blocks.

noahfalk replied:
I assume you mean add PRECONDITION inside EventPipeFile::WriteSequencePoint... Done : )
I didn't add if(pSequencePoint == nullptr) because that is a runtime bug if it ever happens and I'd like to get the AV so we find and fix it quickly.

jorive commented on src/vm/eventpipebuffermanager.cpp:
```diff
@@ -10,6 +10,13 @@
 
 #ifdef FEATURE_PERFTRACING
 
+#ifndef MIN
```
nit: Ideally we should use `std::min` and `std::max` (maybe not available in clr)

noahfalk replied:
There is a macro USE_STL, but it is only defined in the PAL. Elsewhere the runtime avoids STL usage. This avoidance may be historical and no longer have a sound basis, but digging into that investigation probably isn't worth it solely for simple functions like min and max. Looking around I did find a Min/Max implemented in clrtypes.h so I got rid of these defines and converged on those definitions instead.

jorive commented on src/vm/eventpipebuffermanager.cpp:
```diff
@@ -160,6 +172,12 @@ void EventPipeThread::SetBufferList(EventPipeBufferManager *pBufferManager, Even
     EX_END_CATCH(SwallowAllExceptions);
 }
 
+SIZE_T EventPipeThread::GetOSThreadId()
```
nit: const method

jorive commented on src/vm/eventpipebuffermanager.cpp:
```diff
@@ -202,6 +220,25 @@ EventPipeBufferManager::EventPipeBufferManager()
+    m_pCurrentBuffer = nullptr;
+    m_pCurrentBufferList = nullptr;
+
+    m_maxSizeOfAllBuffers = MIN(MAX(100 * 1024, pSession->GetCircularBufferSize()), (ULONGLONG)10 * 1024 * 1024 * 1024);
```
nit: `MIN(MAX(` pattern could be converted to a Clamp function.

jorive replied:
nit: CircularBufferSize could/should be moved to the EventPipeBufferManager

jorive commented on src/vm/eventpipebuffermanager.cpp:
```diff
@@ -202,6 +220,25 @@ EventPipeBufferManager::EventPipeBufferManager()
+
+    m_maxSizeOfAllBuffers = MIN(MAX(100 * 1024, pSession->GetCircularBufferSize()), (ULONGLONG)10 * 1024 * 1024 * 1024);
+
+    if (pSession->GetSequencePointAllocationBudget() == 0)
```
nit: `SequencePointAllocationBudget` could be passed as parameter, and you would not need to pass the `EventPipeSession` object.

noahfalk replied:
In later parts of the change I capture it as a member field and use it in other functions.

noahfalk commented:
I fixed up Linux build issues, found and fixed several bugs, updated for PR comments, and rebased on top of @jorive and @cshung's recent changes. Fingers crossed for CI...

jorive commented on src/vm/eventpipeconfiguration.h:
```diff
@@ -89,16 +89,17 @@ class EventPipeConfiguration
     {
         LIMITED_METHOD_CONTRACT;
-
+        _ASSERTE(EventPipe::MaxNumberOfSessions == 64);
```
nit: Prefer compile time error. Both numbers are `const`, this assert could be a `static_assert`.

jorive commented on src/vm/eventpipesession.cpp:
```diff
@@ -33,6 +34,8 @@ EventPipeSession::EventPipeSession(
         GC_TRIGGERS;
         MODE_PREEMPTIVE;
+        PRECONDITION(index < EventPipe::MaxNumberOfSessions);
+        PRECONDITION(EventPipe::MaxNumberOfSessions == 64); // If MaxNumberOfSessions ever changed, fix the m_id calculation above
```
nit: Same as above w.r.t. using `static_assert` here instead.

jorive commented on src/vm/eventpipethread.h:
```diff
@@ -0,0 +1,151 @@
+
+class EventPipeThreadSessionState
+{
+    // immutable
```
nit: If immutable, please make them `const`

jorive commented on src/vm/eventpipethread.cpp:
```diff
@@ -0,0 +1,235 @@
+    pThread->AddRef();
+}
+
+#ifndef __GNUC__
```
nit: Use the macro `EVENTPIPE_THREAD_LOCAL`

jorive commented on src/vm/CMakeLists.txt:
```diff
@@ -331,6 +331,7 @@ set(VM_SOURCES_WKS
     eventpipemetadatagenerator.cpp
     eventpipeprotocolhelper.cpp
     eventpipeprovider.cpp
+    eventpipethread.cpp
```
Do you need to add the `eventpipethread.h` below (eventpipe*.h headers are there)?

jorive commented on src/vm/eventpipeblock.cpp:
```diff
@@ -57,69 +100,322 @@ bool EventPipeBlock::WriteEvent(EventPipeEventInstance &instance)
+    m_maxTimeStamp.QuadPart = LLONG_MIN;
+}
+
+void WriteVarUInt32(BYTE* & pWritePointer, unsigned int value)
```
nit: Can we make `WriteVarUInt32` and `WriteVarUInt64` a single template function?

noahfalk commented:
I'm going to file an issue for the additional PR feedback you just added and check this in as-is (all the remaining comments are stylistic, no functional bugs). I want to expedite letting other folks test with these changes and I don't want to keep doing rebase+retest as other EventPipe changes come in.
